pkg/kinds/dashboard/validator.go

```diff
@@ -0,0 +1,82 @@
+package dashboard
+
+import (
+	"context"
+	_ "embed"
+	json "encoding/json"
+	fmt "fmt"
+	"strings"
+	"sync"
+
+	"cuelang.org/go/cue"
+	"cuelang.org/go/cue/cuecontext"
+	"cuelang.org/go/cue/errors"
+	cuejson "cuelang.org/go/encoding/json"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+
+	common "github.com/grafana/grafana/pkg/apimachinery/apis/common/v0alpha1"
+)
+
+// ValidateDashboardSpec validates a JSON object against the dashboard kind schema.
+func ValidateDashboardSpec(ctx context.Context, obj common.Unstructured) field.ErrorList {
+	data, err := json.Marshal(obj.Object)
+	if err != nil {
+		return field.ErrorList{
+			field.Invalid(field.NewPath("spec"), field.OmitValueType{}, err.Error()),
+		}
+	}
+
+	if err := cuejson.Validate(data, getSchema()); err != nil {
+		errs := field.ErrorList{}
+
+		for _, e := range errors.Errors(err) {
+			if
+			// We don't want to return confusing "empty disjunction" errors,
+			// because the users don't necessarily understand what to do with them.
+			// For empty disjunctions, CUE will also return more specific errors,
+			// so we can safely ignore the generic ones.
+			strings.Contains(e.Error(), "disjunction") ||
+				// We don't want to return errors about unknown fields either.
+				strings.Contains(e.Error(), "field not allowed") {
+				continue
+			}
+
+			// We want to manually format the error message,
+			// because e.Error() contains the full CUE path.
+			format, args := e.Msg()
+
+			errs = append(errs, field.Invalid(
+				field.NewPath(formatErrorPath(e.Path())),
+				field.OmitValueType{},
+				fmt.Sprintf(format, args...),
+			))
+		}
+
+		return errs
+	}
+
+	return nil
+}
+
+func formatErrorPath(path []string) string {
+	return strings.Join(path[4:], ".")
+}
+
+var (
+	schema        cue.Value
+	getSchemaOnce sync.Once
+)
+
+//go:embed dashboard_kind.cue
+var schemaSource string
```

Comment:
I ended up copy-pasting `dashboard_kind.cue` from `kinds/dashboards` because embedding only works for relative paths and disallows using `.` and `..`.

I think we'll need a more sophisticated long-term solution here.

@spinillos @K-Phoen do you have any thoughts on how we could achieve this? A new custom Jenny in `kinds/generate.go`?

Reply:
@spinillos knows how `kinds/generate.go` works much more than I do, so I'll defer to her. But maybe we could _just_ move the schemas in `kinds/dashboards` to somewhere more convenient to you and adjust the codegen pipeline accordingly?

Slight tangent here: I'm also wondering about the mid/long term future of this pipeline anyway, with the dashboard v2 schema and app-sdk taking over a lot of codegen duties. I imagine that there is a world not too far away in the future where we won't need the `kinds/` folder and its pipeline anymore?

Reply:
> But maybe we could just move the schemas in kinds/dashboards to somewhere more convenient to you and adjust the codegen pipeline accordingly?

That would be the easiest solution!

> Slight tangent here: I'm also wondering about the mid/long term future of this pipeline anyway, with the dashboard v2 schema and app-sdk taking over a lot of codegen duties. I imagine that there is a world not too far away in the future where we won't need the kinds/ folder and its pipeline anymore?

Yeah, that'd be my guess.